xla/service/cpu/onednn_ops_rewriter.cc

```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
+
+  // Match Z = scale / sqrt(variance + eps)
+  if (!Match(scaled_norm_a,
+             m::Multiply().WithBinaryOperandsAnyOrder(
+                 m::Op(&scale_node),
+                 m::Broadcast(
+                     m::Reshape(m::Rsqrt(m::Add().WithBinaryOperandsAnyOrder(
+                         m::Broadcast(m::ConstantScalar(&epsilon)),
+                         m::Op(&sqrd_diff_mean)))))))) {
+    return std::nullopt;
+  }
+
+  // get epsilon
+  *eps = static_cast<float>(epsilon->literal().GetAsDouble({}).value());
+  // get scale
+  if (!Match(scale_node, m::Broadcast(m::Op(scale)))) return std::nullopt;
+
+  // match variance
+  if (!Match(sqrd_diff_mean, MeanPattern(&sqrd_diff))) return std::nullopt;
+
+  if (!Match(sqrd_diff, Square(m::Subtract(
+                            m::Op(&src_a),
+                            m::Broadcast(m::Reshape(MeanPattern(&src_b))))))) {
+    return std::nullopt;
+  }
+
+  if (src_a != src_b && src_a != *src) return std::nullopt;
+
+  // Match mean from Bias - Mean(X)*Z
+  if (!Match(mean0_a, MeanPattern(&src_c))) return std::nullopt;
+
+  if (src_c != *src) return std::nullopt;
+
+  return true;
+}
+
+bool MatchFlaxLayerNorm(HloInstruction* instr, HloInstruction** src,
+                        HloInstruction** scale, HloInstruction** bias,
+                        float* eps, bool* is_bf16orfp16_convert,
+                        bool* is_producer_bf16orfp16,
+                        HloInstruction** convert_instr) {
+  HloInstruction *prod_s, *hinge;
+  HloInstruction *div0, *div1, *div_red;
+  HloInstruction *mul_in0, *mul_in1, *main_pipe_mul_in0;
+  HloInstruction *reduce_in0, *epsilon;
+  HloInstruction *broadcast0, *broadcast1;
+
+  bool scaleFound = false;
+  bool shiftFound = false;
+
+  auto spine = m::Add().WithBinaryOperandsAnyOrder(
+      m::Broadcast(),
+      m::Multiply()
+          .WithBinaryOperandsAnyOrder(
+              m::Op(&hinge).WithOneUser(),
+              m::Subtract(
+                  OptionalConvert(m::Op(&prod_s)),
+                  m::Broadcast(
+                      m::Reshape(
+                          m::Broadcast(m::Reshape(m::Op(&div_red).WithOpcode(
+                                                      HloOpcode::kDivide))
+                                           .WithOneUser())
+                              .WithOneUser())
+                          .WithOneUser())
+                      .WithOneUser())
+                  .WithOneUser())
+          .WithOneUser());
+
+  if (!Match(instr, spine)) return false;
+
+  const Shape& prod_shape = prod_s->shape();
+  if (!IsSupportedType(prod_shape.element_type())) return false;
+
+  HloInstruction* shift = FindLayerNormShift(instr);
+  shiftFound = (shift != nullptr);
+
+  HloInstruction* scale_gamma = FindLayerNormScale(hinge);
+  scaleFound = (scale_gamma != nullptr);
+
+  // Currently patterns without scale and shift are
+  // not supported.
+  // OneDNN only supports 2 <= rank <= 5
```

Comment:
Nit: We can fit the first two lines in one line.
```suggestion
  // Currently patterns without scale and shift are not supported.
  // OneDNN only supports 2 <= rank <= 5
```